README.md
# rightscale_volume cookbook

[![Cookbook](https://img.shields.io/cookbook/v/rightscale_volume.svg?style=flat)][cookbook]
[![Release](https://img.shields.io/github/release/rightscale-cookbooks/rightscale_volume.svg?style=flat)][release]
[![Build Status](https://img.shields.io/travis/rightscale-cookbooks/rightscale_volume.svg?style=flat)][travis]

[cookbook]: https://supermarket.getchef.com/cookbooks/rightscale_volume
[release]: https://github.com/rightscale-cookbooks/rightscale_volume/releases/latest
[travis]: https://travis-ci.org/rightscale-cookbooks/rightscale_volume

# Description

This cookbook provides a rightscale_volume resource that can create, attach and manage a single
block level storage "volume" on numerous public and private IaaS clouds.

A volume provides a highly reliable, efficient storage solution that can be mounted to a
cloud server (within the same datacenter / zone) and persists independently from the life of the instance.

By using the RightScale API, this resource gives your recipes cloud portability without the need
to store your cloud credentials on each server.

Github Repository: [https://github.com/rightscale-cookbooks/rightscale_volume](https://github.com/rightscale-cookbooks/rightscale_volume)

# Requirements

* The system being configured must be a RightScale managed VM to have the required access to the RightScale API.
* Chef 12.
* [RightLink 10](http://docs.rightscale.com/rl10/) See cookbook version 1.2.6 for RightLink 6 support
* Also requires a RightScale account that is registered with all the cloud vendors
  you expect to provision on (e.g. AWS, Rackspace, Openstack, CloudStack, GCE, and Azure).


# Usage

The resource only handles manipulating the volume. Additional resources need to be created in
the recipe to manage the attached volume as a filesystem or logical volume.

The following example will create a 10 GB volume, attach it to the instance, formats the device as ext4
and mounts it to '/mnt/storage'.

```ruby
# Creates a 10 GB volume
rightscale_volume "db_data_volume" do
  size 10
  action :create
end

# Attaches the volume to the instance
rightscale_volume "db_data_volume" do
  action :attach
end

execute "format volume as ext4" do
  command lazy { "mkfs.ext4 #{node['rightscale_volume']['db_data_volume']['device']}" }
  action :run
end

execute "mount volume to /mnt/storage" do
  command lazy { "mkdir -p /mnt/storage; mount #{node['rightscale_volume']['db_data_volume']['device']} /mnt/storage" }
  action :run
end
```

The following example will create a new volume from a snapshot.

```ruby
rightscale_volume "db_data_volume_from_snapshot" do
  size 10
  snapshot_id "my-snaphot-id"
  action [ :create, :attach ]
end
```

The `size` may or may not be honored depending on hypervisor used by the cloud vendor.
If the cloud does not support resize when creating a volume from a snapshot, then the size will be
the same as the volume from which the snapshot was taken. If resize is supported, additional
resources will be required to resize the filesystem on the volume.


# Recipes

## default

The default recipe installs the `right_api_client` RubyGem, which this cookbook requires in
order to work with the RightScale API.


# Resource/Providers

## rightscale_volume

A resource to create, attach, and manage a single "volume" on public and private IaaS clouds.

### Actions

| Name | Description | Default |
| --- | --- | --- |
| `:create` | Creates a new volume in the cloud | yes |
| `:attach` | Attaches a volume to a RightScale server | |
| `:snapshot` | Takes a snapshot of a volume | |
| `:detach` | Detaches a volume from a RightScale server | |
| `:delete` | Deletes a volume from the cloud | |
| `:cleanup` | Cleans up old snapshots of a volume | |

### Attributes

| Name | Description | Default | Required |
| --- | --- | --- | --- |
| `nickname` | Name of the volume to be created | | No |
| `size` | Volume size in gigabytes | `1` | No |
| `description` | Description for the volume | | No |
| `snapshot_id` | Snapshot ID to create the volume from | | No |
| `options` | Optional parameters hash for volume creation. For example, `:volume_type` on Rackspace Open Clouds and `:iops` on AWS clouds | `{}` | No |
| `timeout` | Throws an error if an action could not be completed within this timeout (in minutes) | `15` | No |
| `max_snapshots` | The number of snapshots of a volume to retain when running the `:cleanup` action | `60` | No |


# Cloud Specific Notes

## AWS EC2

* For this resource to work on a EC2 cloud, the RightScale account must be on a
  [UCP](http://support.rightscale.com/12-Guides/Dashboard_Users_Guide/Unified_Cloud_Platform) cluster.
* This cloud supports creating volumes with provisioned IOPS. To create a volume with IOPS on EC2
  pass the `:iops` option to the `options` hash as shown below

```ruby
rightscale_volume "volume_with_iops" do
  size 10
  options {:iops => 100}
  action :create
end
```

## Rackspace Open Cloud

* The minimum volume size offered by this cloud is 100 GB. The `:create` volume action throws an
  error if the requested volume size is lesser than the minimum size offered.
* This cloud supports two types of volume - SATA and SSD. The type of volume to be created can be
  passed to the `options` parameter as below (defaults to SATA if none specified)

```ruby
rightscale_volume "open_cloud_volume" do
  size 100
  options {:volume_type => 'SSD'}
  action :create
end
```
* A volume cannot be deleted from this cloud if at least one snapshot created from this volume
  exists. To delete such a volume, all dependent snapshots must be cleaned up first. The `:delete`
  action does not delete such a volume and throws a warning message in the logs.

## CloudStack Clouds

* CloudStack has the concept of a "custom" disk offering. If a "custom volume type" is supported in the cloud,
  then the `:create` action creates a volume with the requested size. If "custom volume type" is not supported
  then this action will use the "closest volume type" with size greater than or equal to the requested size.
  If there are multiple custom volume types or multiple volume types with the closest size, the one with the greatest
  resource UID will be used.

# Author

Author:: RightScale, Inc. (<cookbooks@rightscale.com>)


CHANGELOG.md
rightscale_volume Cookbook CHANGELOG
====================================

This file is used to list changes made in each version of the rightscale_volume cookbook.

v2.0.0
------
- Updated for Chef 12
- Updated for ChefDK 1.0.3

v1.3.1
------
- Pin right_api_client at v1.6.1 for ruby 1.9.3 support, Chef 11

v1.2.10
-------

- On CloudStack KVM, do not delete volumes as backups depend on them to restore.

v1.2.9
------

- Verify from OS perspective that volume has been attached.

v1.2.8
------

- Use correct Rackspace cloud name.

v1.2.7
------

- Include 'build-essentials' which installs packages required for compiling and installing right_api_client.

v1.2.6
------

- On AWS, determine next volume device name based on AWS UserGuide.

v1.2.5
------

- Don't pass volume type when restoring with snapshot ID on Cloudstack.

v1.2.4
------

- Scanning of volume changes to be done on VMware; previously only done on vSphere.

v1.2.3
------

- Support volume types on all clouds instead of just those that require it ([#16][]).

v1.2.2
------

- Add 20 minute timeout for right_api_client calls.

v1.2.1
------

- `controller_type` can be passed in as an option.

v1.2.0
------

- Add [shields.io](http://shields.io/) badges to README.
- Add VMware vSphere support.

v1.1.1
------

- [#9][]: Do not raise from load_current_resource when a volume that previously existed has been removed.

v1.1.0
------

- A `nickname` attribute is added which defaults to the value given in name attribute. This is useful if there are
  multiple resource actions on the same resource and the name needs to be unique.
- The `:detach` action will not raise an exception if the volume is not found.
- Fixed a bug where the volume was not created from the snapshot properly.
- The exception raised when a volume is deleted on OpenStack where it has dependent snapshots is rescued and logged.
- AWS Provisioned IOPS is supported on volume creation.
- Removed unused `delayed_evaluator` cookbook now that the `lazy` evaluator is included in Chef 11.

v1.0.1
------

- Pull Request [#6][]: Also check for 'gce' cloud type for google
- Added ChefSpec matchers

v1.0.0
------

- Initial release

<!--- The following link definition list is generated by PimpMyChangelog --->
[#6]: https://github.com/rightscale-cookbooks/rightscale_volume/issues/6
[#9]: https://github.com/rightscale-cookbooks/rightscale_volume/issues/9
[#16]: https://github.com/rightscale-cookbooks/rightscale_volume/issues/16


recipes/test.rb
#
# Cookbook Name::network_storage
#
# Copyright RightScale, Inc. All rights reserved. All access and use subject to the
# RightScale Terms of Service available at http://www.rightscale.com/terms.php and,
# if applicable, other agreements such as a RightScale Master Subscription Agreement.

# Test configurations
actions_to_test = [:create, :attach, :snapshot, :cleanup, :detach, :delete]
# To test the volume creation from a snapshot
snap_id = nil
# To test IOPS on AWS clouds
test_iops = nil

# Set the provider based on the cloud
#if node[:cloud][:provider] == 'ec2'
#  provider_klass = Chef::Provider::NetworkStorageDeviceEC2
#else
#  provider_klass = Chef::Provider::NetworkStorageDeviceAPI15
#end

Chef::Log.info "  ======= NETWORK STORAGE DEVICE TESTER ========"

Chef::Log.info "  **********************************************"
Chef::Log.info "  ********* Actions: Create & Attach ***********"
Chef::Log.info "  **********************************************"
rightscale_volume "test_device" do
  size 10
  description "test block device created inside vagrant"
  snapshot_id snap_id if snap_id
  if test_iops
    options(
      :iops => test_iops
    )
  end
  action [:create, :attach]
  only_if { actions_to_test.include?(:create) && actions_to_test.include?(:attach) }
end

Chef::Log.info "  **********************************************"
Chef::Log.info "  ************* Actions: Snapshot **************"
Chef::Log.info "  **********************************************"
rightscale_volume "test_device" do
  action :snapshot
  only_if { actions_to_test.include?(:snapshot) }
end

Chef::Log.info "  **********************************************"
Chef::Log.info "  ************* Actions: Cleanup ***************"
Chef::Log.info "  **********************************************"
rightscale_volume "test_device" do
  action :cleanup
  only_if { actions_to_test.include?(:cleanup) }
end

Chef::Log.info "  **********************************************"
Chef::Log.info "  ************* Actions: Detach ****************"
Chef::Log.info "  **********************************************"
rightscale_volume "test_device" do
  action :detach
  only_if { actions_to_test.include?(:detach) }
end

Chef::Log.info "  **********************************************"
Chef::Log.info "  ************** Actions: Delete ***************"
Chef::Log.info "  **********************************************"
rightscale_volume "test_device" do
  action :delete
  only_if { actions_to_test.include?(:delete) }
end

Chef::Log.info "  All actions are executed and finished successfully"


spec/spec_helper.rb
#
# Cookbook Name:: rightscale_volume
# Spec:: spec_helper
#
# Copyright (C) 2013 RightScale, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

lib = File.expand_path('../../libraries', __FILE__)
$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)

require 'chefspec'
require 'provider_rightscale_volume'
require 'resource_rightscale_volume'
ChefSpec::Coverage.start!


recipes/default.rb
#
# Cookbook Name:: rightscale_volume
# Recipe:: default
#
# Copyright (C) 2013 RightScale, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

log 'Installing required gems and dependencies...'

# Install build-essentials at compile time so it is available for right_api_client
node.normal['build-essential']['compile_time'] = true
include_recipe 'build-essential'

# Install gems during compile phase so that they are available to files
# which require them during converge phase.
chef_gem 'right_api_client' do
  compile_time true
end


libraries/matchers.rb
#
# Cookbook Name:: rightscale_volume
# Library:: matchers
#
# Copyright (C) 2014 RightScale, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if defined?(ChefSpec)
  def create_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :create, resource_name)
  end

  def attach_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :attach, resource_name)
  end

  def detach_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :detach, resource_name)
  end

  def delete_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :delete, resource_name)
  end

  def snapshot_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :snapshot, resource_name)
  end

  def cleanup_rightscale_volume(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new('rightscale_volume', :cleanup, resource_name)
  end
end


libraries/resource_rightscale_volume.rb
#
# Cookbook Name:: rightscale_volume
# Library:: resource_rightscale_volume
#
# Copyright (C) 2013 RightScale, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/resource'
require 'chef/resource/lwrp_base'

class Chef
  class Resource
    # A resource class for 'rightscale_volume' cookbook.
    class RightscaleVolume < Chef::Resource::LWRPBase
      provides :rightscale_volume
      resource_name :rightscale_volume

      # Volume state
      attr_accessor :state

      # Device to which the volume is attached
      attr_accessor :device

      # Initializes rightscale_volume resource.
      #
      # @param name [String] name of the resource.
      # @param run_context [Chef::RunContext] optional value to track context of
      # a chef run.
      # @return [Chef::Resource::RightscaleVolume] the newly created
      # rightscale_volume resource.
      #
      # def initialize(name, run_context = nil)
      #  super
      #  @resource_name = :rightscale_volume
      #  @action = 'create'
      #  @allowed_actions.push(:create, :delete, :attach, :detach, :snapshot, :cleanup)
      #  @provider = Chef::Provider::RightscaleVolume
      # end
      actions :create, :delete, :attach, :detach, :snapshot, :cleanup
      default_action :create

      # Name of the device.
      #
      # @param arg [String] the volume name
      #
      # @return [String] the volume name
      #
      # def nickname(arg = nil)
      #  set_or_return(
      #    :name,
      #    arg,
      #    kind_of: String
      #  )
      # end
      attribute :nickname, kind_of: String, name_attribute: true, required: true

      # Size of the device to create in GBs. By default, this is set to 1 GB.
      #
      # @param arg [Integer] the volume size
      #
      # @return [Integer] the volume size
      #
      # def size(arg = nil)
      #  set_or_return(
      #    :size,
      #    arg,
      #    kind_of: Integer,
      #    default: 1
      #  )
      # end
      attribute :size, kind_of: Integer, default: 1

      # Description of the device.
      #
      # @param arg [String] the volume description
      #
      # @return [String] the volume description
      #
      # def description(arg = nil)
      #  set_or_return(
      #    :description,
      #    arg,
      #    kind_of: String
      #  )
      # end
      attribute :description, kind_of: String

      # Identifier for a single volume.
      #
      # @param arg [String] the volume ID
      #
      # @return [String] the volume ID
      #
      # def volume_id(arg = nil)
      #  set_or_return(
      #    :volume_id,
      #    arg,
      #    kind_of: String
      #  )
      # end
      attribute :volume_id, kind_of: String

      # Name of snapshot.
      #
      # @param arg [String] the snapshot name.
      #
      # @return [String] the snapshot name.
      #
      # def snapshot_name(arg = nil)
      #   set_or_return(
      #     :snapshot_name,
      #     arg,
      #     kind_of: String
      #   )
      # end
      attribute :snapshot_name, kind_of: String

      # Snapshot ID to create the volume from.
      #
      # @param arg [String] the snapshot ID
      #
      # @return [String] snapshot ID
      #
      # def snapshot_id(arg = nil)
      #   set_or_return(
      #     :snapshot_id,
      #     arg,
      #    kind_of: String
      #   )
      # end
      attribute :snapshot_id, kind_of: String

      # Maximum number of snapshots to keep. This is used during cleanup.
      # By default, this is set to 60.
      #
      # @param arg [Integer] the number of snapshots to retain when running the
      # +:cleanup+ action
      #
      # @return [Integer] the number of snapshots to retain
      #
      # def max_snapshots(arg = nil)
      #   set_or_return(
      #     :max_snapshots,
      #     arg,
      #     kind_of: Integer,
      #     default: 60
      #   )
      # end
      attribute :max_snapshots, kind_of: Integer, default: 60

      # Timeout in minutes. Used to throw an error if the action cannot be
      # completed by the cloud provider within this given minutes. By default,
      # this is set to 15 minutes.
      #
      # @param arg [Integer] the timeout value in minutes for actions
      #
      # @return [Integer] the timeout value
      #
      # def timeout(arg = nil)
      #   set_or_return(
      #    :timeout,
      #    arg,
      #    kind_of: Integer,
      #    default: 15
      #  )
      # end
      attribute :timeout, kind_of: Integer, default: 15

      # Hash that holds cloud provider specific attributes such as IOPS or
      # volume_type ('SATA'/'SSD').
      #
      # @param arg [Hash] the optional parameters for actions
      #
      # @return [Hash] the optional parameters
      #
      # def options(arg = nil)
      #  set_or_return(
      #    :options,
      #    arg,
      #    kind_of: Hash,
      #    default: {}
      #  )
      # end
      attribute :options, kind_of: Hash, default: {}
    end
  end
end
